### PR TITLE
Run encoder on IPU for encoder-decoder text-gen models

### PR DESCRIPTION
**Pull request by jimypbr**
# What does this PR do?

Optimisation for encoder/decoder text-gen models that runs the encoder part of the model on IPU instead of CPU.
It does this by automatically splitting the IPUConfig into two different configs for the encoder and decoder based upon the `ipu_config.layers_per_ipu` and the corresponding `num_encoder_layers` and `num_decoder_layers` from the model.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/modeling_utils.py [hunk 1]
```diff
@@ -225,6 +225,32 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
...
+        "'{arg}' must take one of the following three forms:\n"
+        "  - float: one value for all parts\n"
+        "  - list(float): different values for each parts\n"
+        "  - list(list(float)): different values for each IPU in each part\n"
+        "The outer list must be of length {n}"
+    )
+    # Split layers_per_ipu between the given parts
+    ipu_configs = {part: copy.deepcopy(ipu_config) for part in parts}
+    for i, (name, ipu_config) in enumerate(ipu_configs.items()):
+        if isinstance(ipu_config.layers_per_ipu, list):
```
[the remark this answers is not in the document]
Comment (HMellor) [reply optimum/graphcore/modeling_utils.py:239]: If the user passes an integer, do we put it into a list?

File: optimum/graphcore/modeling_utils.py [hunk 2]
```diff
@@ -225,6 +225,32 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
             return sum(p.numel() for p in self.parameters() if p.requires_grad or not only_trainable)
 
 
+def split_ipu_config(ipu_config: IPUConfig, parts: List[str] = []) -> List[IPUConfig]:
+    msg = (
```
[the remark this answers is not in the document]
Comment (HMellor) [reply optimum/graphcore/modeling_utils.py:229]: Is this comment in the right place? I'm not sure I follow

File: tests/test_ipu_configuration.py [hunk 3]
```diff
@@ -304,3 +304,53 @@ def test_layers_per_ipu(self):
         ipu_config = IPUConfig(ipus_per_replica=0)
         with pytest.raises(ValueError, match=r"Invalid value for ipus_per_replica"):
             layer_ipu = get_layer_ipu(ipu_config, 6)
+
+    def test_split_encoder_decoder_ipu_config(self):
```
Comment (HMellor) [inline tests/test_ipu_configuration.py:308]: All the tests are good, less keen that they are all in one monolithic test.

Splitting them up and [grouping them in a class](https://docs.pytest.org/en/7.1.x/getting-started.html#group-multiple-tests-in-a-class) would be good.

File: optimum/graphcore/generation_utils.py [hunk 4]
```diff
@@ -80,45 +100,49 @@ def _pad_tensors_to_max_len(self, tensor: torch.Tensor, max_length: int, pad_tok
...
         encoder_kwargs[model_input_name] = inputs_tensor
-        model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)
-
-        # 5. undo the casting for cpu
-        if dtype is torch.float16:
-            encoder.to(torch.float16)
-            for key, output in model_kwargs["encoder_outputs"].items():
-                if isinstance(output, torch.Tensor) and output.dtype is torch.float32:
-                    model_kwargs["encoder_outputs"][key] = output.to(torch.float16)
+        if not hasattr(self, "poptorch_encoder"):
```
Comment (HMellor) [inline optimum/graphcore/generation_utils.py:130]: Can you either use this in both places or use `x_ipu_config = getattr(self, "x_ipu_config", self.ipu_config)` in both places like you did for the decoder?

File: optimum/graphcore/generation_utils.py [hunk 5]
```diff
@@ -12,7 +12,12 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
+import contextlib
+import copy
```
Comment (HMellor) [inline optimum/graphcore/generation_utils.py:16]: Doesn't look like this is used based on the diff

File: optimum/graphcore/generation_utils.py [hunk 6]
```diff
@@ -12,7 +12,12 @@
...
+import json
+import os
 import warnings
+from itertools import groupby
```
Comment (HMellor) [inline optimum/graphcore/generation_utils.py:20]: No longer used

File: optimum/graphcore/generation_utils.py [hunk 7]
```diff
@@ -80,45 +100,49 @@ def _pad_tensors_to_max_len(self, tensor: torch.Tensor, max_length: int, pad_tok
...
+            # Use split encoder ipu_config for encoder/decoder models
+            self.poptorch_encoder = poptorch.inferenceModel(
+                encoder.eval(), self.encoder_ipu_config.to_options(for_inference=True)
+            )
+        with graph_profile_dir_append("/encoder"):
+            model_kwargs["encoder_outputs"]: ModelOutput = self.poptorch_encoder(**encoder_kwargs)
 
         return model_kwargs
 
+    def detachFromDevice(self):
```
Comment (HMellor) [inline optimum/graphcore/generation_utils.py:140]: Will this cause problems if a pipeline is created with a non split model? Or is this class only used for split models?

File: optimum/graphcore/modeling_utils.py [hunk 8]
```diff
@@ -241,7 +253,7 @@ def get_layer_ipu(ipu_config: IPUConfig, target_number_of_layers: Optional[Union
 
         # if ipus_per_replica is 1, then put everything on IPU0, ignoring layers_per_ipu
         if ipus_per_replica == 1:
-            return [0] * target_number_of_layers
+            return [target_number_of_layers]
```
Comment (HMellor) [inline optimum/graphcore/modeling_utils.py:256]: Good catch

File: optimum/graphcore/modeling_utils.py [hunk 9]
```diff
@@ -278,6 +290,76 @@ def get_layer_ipu(ipu_config: IPUConfig, target_number_of_layers: Optional[Union
...
+    """
+    # Need at least two IPUs to do the split
+    if ipu_config.ipus_per_replica < 2:
+        raise ValueError("Need ipus_per_replica of at least 2 to split ipu_config into encoder and decoder configs")
+
+    ipu_configs = {name: copy.deepcopy(ipu_config) for name in ["encoder", "decoder"]}
+
+    # Split layers_per_ipu between the given num layers
+    layers_per_ipu = _expand_layers_per_ipu_wildcard(ipu_config, num_encoder_layers + num_decoder_layers)
+    cumsum = [sum(layers_per_ipu[: i + 1]) for i in range(len(layers_per_ipu))]
```
Comment (HMellor) [inline optimum/graphcore/modeling_utils.py:334]: What happens if there's a `0` at the cut?

i.e. if we had `[1, 2, 3, 0, 5, 6, 7, 8]` would it get split into `[1, 2, 3]` and `[0, 5, 6, 7, 8]`?

**Review by HMellor — CHANGES_REQUESTED** [review]

File: optimum/graphcore/generation_utils.py (hunk 4 above)
Comment (jimypbr) [reply optimum/graphcore/generation_utils.py:130]: The decoder one is a special case that needs that check because models like GPT are decoder-only

File: optimum/graphcore/generation_utils.py (hunk 7 above)
Comment (jimypbr) [reply optimum/graphcore/generation_utils.py:140]: This might cause problems if the model is being used like in training mode in the `IPUTrainer`.

File: optimum/graphcore/modeling_utils.py (hunk 9 above)
Comment (jimypbr) [reply optimum/graphcore/modeling_utils.py:334]: Good catch!

File: tests/test_ipu_configuration.py (hunk 3 above)
Comment (HMellor) [reply tests/test_ipu_configuration.py:308]: I think I'll do this in a separate PR to make this change on other tests too (if I feel it's worth it)

File: optimum/graphcore/generation_utils.py (hunk 7 above)
Comment (HMellor) [reply optimum/graphcore/generation_utils.py:138]: I suppose if the model isn't split it'll just be a no-op, but we'd like it to work for all possible poplar executors that can exist in generation

File: optimum/graphcore/pipelines/__init__.py [hunk 10]
```diff
@@ -226,7 +226,10 @@ def get_poplar_executor(
     ipu_config.inference_replication_factor = 1
     try:
         model = to_pipelined(model, ipu_config, force=False)
-        model.parallelize()
+        if model.config.is_encoder_decoder:
+            model.parallelize(for_generation=True)
+        else:
+            model.parallelize()
```
[the remark this answers is not in the document]
Comment (jimypbr) [reply optimum/graphcore/pipelines/__init__.py:232]: No the `is_encoder_decoder` is a taxonomy of the model and doesn't change at run time. We need a `for_generation` so we can run encoder-decoder models in either "training mode" (encoder and decoder in same executor) or in generation mode.

File: optimum/graphcore/generation_utils.py [hunk 11]
```diff
@@ -42,10 +45,25 @@
 from transformers.modeling_outputs import ModelOutput
 from transformers.pytorch_utils import torch_int_div
 
+from .modeling_utils import get_layer_ipu
```
Comment (HMellor) [inline optimum/graphcore/generation_utils.py:48]: Not used here?
```suggestion
```

File: optimum/graphcore/generation_utils.py (hunk 7 above)
Comment (HMellor) [reply optimum/graphcore/generation_utils.py:136]: Did you check this?

File: optimum/graphcore/modeling_utils.py [hunk 12]
```diff
@@ -225,23 +231,35 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
...
+    ```
+    this function will expand the wildcard values to `layers_per_ipu = [4, 5]`
+    """
+    layers_per_ipu = copy.deepcopy(ipu_config.layers_per_ipu)
     ipus_per_replica = ipu_config.ipus_per_replica
 
     # Check inputs are valid
     if not all(isinstance(n, int) and n >= -1 for n in layers_per_ipu):
-        raise ValueError("Invalid values in layers_per_ipu. " f"layers_per_ipu={layers_per_ipu}")
+        raise IncompatibleIPUConfigError("Invalid values in layers_per_ipu. " f"layers_per_ipu={layers_per_ipu}")
```
Comment (HMellor) [inline optimum/graphcore/modeling_utils.py:252]:
```suggestion
        raise IncompatibleIPUConfigError(f"Invalid values in layers_per_ipu. layers_per_ipu={layers_per_ipu}")
```
